doc/design/block.md

```diff
@@ -0,0 +1,128 @@
+# Block design
+
+In computer programming, a block is a lexical structure of source code which is grouped together.
+In most programming languages, block is useful when define a function or some conditional statements such as `if` and `while`.
+
+In Paddle, we need a similar concept to support following scenes:
+
+- the basic unit of execution, just like the original `NetOp`, block can group multiple operators and make them behave just like a single operator. Unlike `NetOp`, a network should not be an operator, but a block can be treated as a operator.
+  - Block should have a `Exec` method, which will execute the operators grouped in the block.
+- store the context for operators, including the information of variables they will operator on. 
+- Support the definition and execution of special operators such as `RNNOp`, `switch_op`, `while_op`, `if_else_op`, which needs one or more code-blocks.
+  - Block will help to define and execute thouse operators.
+  
+## when need a block
+Blocks are needed in following code snippets
+```c++
+if(cond) { block_true } else { block_false }
+```
+
+```c++
+while(cond){ block }
+```
+
+```c++
+int func(inputs) { block with outputs }
+```
+
+```c++
+namespace xxxx {
+  block
+}
+```
+
+Paddle will have `if_else_op`, `switch_op`, `while_op` and some other operators that need a function block such as RNNOp, and enable definition and execution of a block is vital to those operators.
+
+What's more, Paddle should has a `pd.namespace`, which will help to preventing name conflicts of variables in large model.
```

Comment:
What's the relationship between namespace and scope?

Reply:
namespace is on the top of a scope, namespace just add a namespace-prefix to the local variable in a scope.

Reply:
If so, we only need one global scope, and we use different namespace to prevent name conflicts. Am I right?

Reply:
Yes. 

Scope is from Caffe2's WorkSpace, op developers is free to use Scope when implement new operators, but we will provide `pd.namespace` to users, and hide the Scope concept from them.

Reply:
@QiJune To what I understand from this design doc draft, the namespace is one of the several uses of Block. Other uses include IfOp, WhileOp, RNNOp. @Superjom am I right?

Reply:
in my understanding, namespace is more widely used then block, for example, we can define a function inside a namespace without block.

Reply:
yes